astrodata/nddata.py

```diff
@@ -5,52 +5,46 @@
 
 from __future__ import (absolute_import, division, print_function)
 
-from copy import deepcopy
 import warnings
+from copy import deepcopy
 
 from astropy.nddata import NDData
-from astropy.nddata import StdDevUncertainty
 from astropy.nddata.mixins.ndslicing import NDSlicingMixin
 from astropy.nddata.mixins.ndarithmetic import NDArithmeticMixin
 from astropy.io.fits import ImageHDU
 
 import numpy as np
 
-__all__ = ['NDAstroData']
-
-
-class StdDevAsVariance(object):
-
-    def as_variance(self):
-
-        if self.array is not None:
-            return self.array ** 2
-
-        else:
-            return None
-
-
-def new_variance_uncertainty_instance(array):
-
-    if array is None:
-        return
+try:
+    from astropy.nddata import (VarianceUncertainty, NDUncertainty,
+                                IncompatibleUncertaintiesException)
+except ImportError:
+    # For Astropy < 3.1
+    from .nduncertainty import (VarianceUncertainty, NDUncertainty,
+                                IncompatibleUncertaintiesException)
 
-    with warnings.catch_warnings():
+__all__ = ['NDAstroData']
 
-        if (array < 0.).any():
-            warnings.warn("Negative variance values found. Setting to zero.",
-                          RuntimeWarning)
-            array = np.where(array >= 0., array, 0.)
 
-        warnings.simplefilter("ignore", RuntimeWarning)
+class ADVarianceUncertainty(VarianceUncertainty):
+    """
+    Subclass VarianceUncertainty to check for negative values.
+    """
 
-        obj = StdDevUncertainty(np.sqrt(array))
+    def __init__(self, array=None, copy=True, unit=None):
+        if array is not None:
+            if isinstance(array, NDUncertainty):
+                if array.uncertainty_type != self.uncertainty_type:
+                    raise IncompatibleUncertaintiesException
+                array = array.array
 
-        warnings.simplefilter("default", RuntimeWarning)
+            if array is not None and np.any(array < 0):
```

Comment:
Good idea, I did not think about it since the original code ust check at instantiation, but this would actually be cleaner since `__init__` uses the setter, and this would avoid the need to check for the uncertainty type, which is done in the parent class. I will give it a try.

Reply:
:+1: (also to the other comments). Of course someone can still change the values of the array to be negative afterwards, but I don't think we should be trying to prevent that anyway -- just to ensure that whenever the user provides an array to be used as variance data, it is suitable for that purpose.

Reply:
Done (168a4e9f614d1fcbabac3fa166c541a5d76c76a7), I will wait for the tests to run and then merge.